Index of edges and vertices in storage layer

bright-starry-sky opened the pull request:
Support for edges and vertices insert、delete and update

bright-starry-sky commented:
Jenkins go

dangleptr commented on src/common/base/NebulaKeyUtils.cpp:
```diff
@@ -8,6 +8,8 @@
 
 namespace nebula {
 
+const std::string kIndexPrefix = "_i_";           // NOLINT
```
Do we still need it?  The first byte could be used as the prefix.

bright-starry-sky replied:
> Do we still need it? The first byte could be used as the prefix.

make sense !  Will using kIndex for prefix.

dangleptr commented on src/common/base/NebulaKeyUtils.cpp:
```diff
@@ -88,6 +90,69 @@ std::string NebulaKeyUtils::kvKey(PartitionID partId, const folly::StringPiece&
+std::string NebulaKeyUtils::vertexIndexKey(PartitionID partId, IndexID indexId, VertexID vId,
+                                           const IndexValues& values) {
+    auto raw = indexRaw(values);
+    int32_t item = (partId << 8) | static_cast<uint32_t>(NebulaKeyType::kData);
```
NebulaKeyType::kIndex ?

bright-starry-sky replied:
> NebulaKeyType::kIndex ?

It's really all set，great！

dangleptr commented on src/common/base/NebulaKeyUtils.cpp:
```diff
@@ -88,6 +90,69 @@ std::string NebulaKeyUtils::kvKey(PartitionID partId, const folly::StringPiece&
+    key.append(reinterpret_cast<const char*>(&item), sizeof(int32_t))
+       .append(kIndexPrefix.data(), kIndexPrefix.size())
+       .append(reinterpret_cast<const char*>(&indexId), sizeof(IndexID))
+       .append(raw.data(), raw.size())
```
Could we avoid the double copy for "raw"?

dangleptr commented on src/interface/storage.thrift:
```diff
@@ -208,15 +218,17 @@ struct EdgeKeyRequest {
+    1: common.GraphSpaceID      space_id,
+    2: common.PartitionID       part_id,
+    3: common.VertexID          vid;
+    4: optional list<IndexItem> indexes,
```
Why we need indexes here?

dangleptr commented on src/interface/storage.thrift:
```diff
@@ -199,6 +207,8 @@ struct AddEdgesRequest {
     // If true, it equals an upsert operation.
     3: bool overwritable,
+    // index item
+    4: optional list<IndexItem> indexes,
```
Why we need the indexes here?  
Storaged could read the information from metad.  We don't need to pass the information every request.

bright-starry-sky replied:
> Why we need the indexes here?
> Storaged could read the information from metad. We don't need to pass the information every request.

Aha, Index meta wasn't available then , let me improve it now.

dangleptr commented on src/interface/storage.thrift:
```diff
@@ -208,15 +218,17 @@ struct EdgeKeyRequest {
     1: common.GraphSpaceID space_id,
     // partId => edgeKeys
     2: map<common.PartitionID, list<EdgeKey>>(cpp.template = "std::unordered_map") parts,
+    3: optional list<IndexItem> indexes,
```
ditto

dangleptr commented on src/interface/storage.thrift:
```diff
@@ -282,23 +294,25 @@ struct UpdateItem {
+    5: list<UpdateItem>         update_items,
+    6: list<binary>             return_columns,
+    7: bool                     insertable,
+    8: optional list<IndexItem> indexes,
```
ditto

dangleptr commented on src/interface/storage.thrift:
```diff
@@ -282,23 +294,25 @@ struct UpdateItem {
+    5: list<UpdateItem>         update_items,
+    6: list<binary>             return_columns,
+    7: bool                     insertable,
+    8: optional list<IndexItem> indexes,
```
ditto

dangleptr commented on src/storage/AddEdgesProcessor.cpp:
```diff
@@ -13,28 +13,116 @@ namespace nebula {
 namespace storage {
 
 void AddEdgesProcessor::process(const cpp2::AddEdgesRequest& req) {
```
asyncAtomicOp will hurt the performance. 
If there is no index,  we should use put interface directly.

critical27 commented on src/storage/test/IndexTest.cpp:
```diff
@@ -0,0 +1,674 @@
+                rowCount++;
+                iter->next();
+            }
+            EXPECT_EQ(10, rowCount);
```
Why we have 10 lines of index with only 9 lines of data?

bright-starry-sky commented:
Addressed all @dangleptr comments since 2019-12-15 10:15:00:
1:  Using kIndex instead of kData.
2: Removed 'kIndexPrefix' from index key.
3: Removed 'IndexItem' from all thrift requests.
4: Avoid the use of `asyncAtomicOp` if no index exists.
5: Avoid the double copy for `NebulaKeyUtils::vertexIndexKey` and `NebulaKeyUtils::edgeIndexKey`.

critical27 commented on src/common/base/NebulaKeyUtils.cpp:
```diff
@@ -88,6 +88,59 @@ std::string NebulaKeyUtils::kvKey(PartitionID partId, const folly::StringPiece&
+    key.append(reinterpret_cast<const char*>(&item), sizeof(int32_t))
+       .append(reinterpret_cast<const char*>(&indexId), sizeof(IndexID));
+    indexRaw(values, key);
+    key.append(reinterpret_cast<const char*>(&srcId), sizeof(VertexID))
```
Why we don't encode edgeType here?

critical27 commented on src/common/base/NebulaKeyUtils.cpp:
```diff
@@ -88,6 +88,59 @@ std::string NebulaKeyUtils::kvKey(PartitionID partId, const folly::StringPiece&
+    key.append(reinterpret_cast<const char*>(&item), sizeof(int32_t))
+       .append(reinterpret_cast<const char*>(&indexId), sizeof(IndexID));
+    indexRaw(values, key);
+    key.append(reinterpret_cast<const char*>(&srcId), sizeof(VertexID))
```
I am sorry for my poor understanding
1. Why we don't need to encode edge type here?
2. Why we encode the `edge key` in index key instead of index value?
3. I guess we can scan all index of first property, am I right?

bright-starry-sky replied:
> I am sorry for my poor understanding
> 
> 1. Why we don't need to encode edge type here?
IndexId can  identify the `edgetype` or `tagId`
> 2. Why we encode the `edge key` in index key instead of index value?
we only need scan key by prefix or range, so we need encode the values into key , I don't know if my answer is clear. 
> 3. I guess we can only scan all index of first property in IndexValues, am I right?
Yes，you are right, if is a range query ,only scan first property. if is a equivalent query, can be first N properties.

critical27 replied:
Maybe the second one need a little more hint.

bright-starry-sky replied:
> Maybe the second one need a little more hint.

for second one, the index key format is :
`partId + indexId + propValues + srcId + rank + dstId . `
the propValues format is : 
`edgeType.col1 + edgeType.col2 + degeType.col3+... `. the column value is encoded(int and timestamp is big-endain).

bright-starry-sky commented on src/storage/AddEdgesProcessor.cpp:
```diff
@@ -13,27 +13,133 @@ namespace nebula {
+                cpp2::EdgeKey ek(apache::thrift::FragileConstructor::FRAGILE,
+                        srcId, type, rank, dstId);
+                auto ekVal = NebulaKeyUtils::edgePrefix(partId, srcId, type, rank, dstId);
+                newIndexes[ekVal] = std::make_pair(edge.get_props(), std::make_pair(ek, index));
```
> Why we encode the edgeKey both in key and value of `newIndexes`? I guess you mean to prevent from different edge of same property value? Maybe `map<propertry, vector<edgekey>>` is enough?

good point ！There are two points here：
1: As your said,  different edge of same property value.
2: Same edge but different version number in a request.

critical27 replied:
Still a little confused, I have not see the place we use the key of `newIndexes` (aka `ekVal`). Besides, `ek` and `ekVal` is almost the same.

As the second case `Same edge but different version number in a request`, we didn't encode any version here. I see you insert a new index in `newIndex` and discards all index of the edges we insert in `obsoleteIndex`.

bright-starry-sky replied:
> Still a little confused, I have not see the place we use the key of `newIndexes` (aka `ekVal`). Besides, `ek` and `ekVal` is almost the same.
> 
> As the second case `Same edge but different version number in a request`, we didn't encode any version here. I see you insert a new index in `newIndex` and discards all index of the edges we insert in `obsoleteIndex`.

For example : 
insert below edges in one request:
kv (part1_src1_edgeType1_rank1_dst1 , v1)
kv (part1_src1_edgeType1_rank1_dst1 , v2)
kv (part1_src1_edgeType1_rank1_dst1 , v3)
kv (part1_src1_edgeType1_rank1_dst1 , v4)

Ultimately, only one record is stored in newIndexes ：kv (part1_src1_edgeType1_rank1_dst1 , v4) . It's just what I need.Answer your question，the `ekVal` only is a unique key.

critical27 replied:
👍

critical27 commented on src/storage/test/IndexTest.cpp:
```diff
@@ -0,0 +1,670 @@
+                rowCount++;
+                iter->next();
+            }
+            EXPECT_EQ(10, rowCount);
```
Why we will have one more index row?

bright-starry-sky replied:
Aha, Let's do the math：
**3part * 10vertex * 9tag = 270(row)**
```
for (auto partId = 0; partId < 3; partId++) {
        for (auto vertexId = 10 * partId; vertexId < 10 * (partId + 1); vertexId++) {
            EXPECT_EQ(9, rowCount);
        }
    }
```

**3part * 9index(same with tag num) * 10 vertex = 270(row)**
```
    for (auto partId = 0; partId < 3; partId++) {
        for (auto indexId = 3001; indexId < 3010; indexId++) {
            EXPECT_EQ(10, rowCount); <<-- this is means 10 vertices.
        }
    }
```

critical27 replied:
My bad, got it.

darionyaphet commented on src/storage/test/DeleteVertexTest.cpp:
```diff
@@ -20,9 +20,10 @@ namespace storage {
 TEST(DeleteVertexTest, SimpleTest) {
     fs::TempDir rootPath("/tmp/DeleteVertexTest.XXXXXX");
     std::unique_ptr<kvstore::KVStore> kv(TestUtils::initKV(rootPath.path()));
+    auto schemaMan = TestUtils::mockSchemaMan();
```
Do we need `schema manager` at here ?

bright-starry-sky replied:
Yes, we need check the existing index via schemaMan.

darionyaphet commented on src/storage/AddVerticesProcessor.cpp:
```diff
@@ -22,30 +22,133 @@ void AddVerticesProcessor::process(const cpp2::AddVerticesRequest& req) {
-    auto spaceId = req.get_space_id();
+    spaceId_ = req.get_space_id();
     callingNum_ = partVertices.size();
+    auto iRet = schemaMan_->getTagIndexes(spaceId_);
```
I have added an index manager maybe you can use it to replace `schemaMan_`. I will release it later.

bright-starry-sky replied:
That's cool !!

bright-starry-sky commented:
Resolve the conflict

critical27 commented on src/storage/AddEdgesProcessor.cpp:
```diff
@@ -13,27 +13,133 @@ namespace nebula {
+        std::for_each(req.parts.begin(), req.parts.end(), [&](auto& partEdges){
+            auto partId = partEdges.first;
+            const auto &edges = partEdges.second;
+            auto atomic = [&]() -> std::string {
```
Maybe you should capture by value here, for example partId could be released.

bright-starry-sky replied:
Sorry, I don't quite follow you. your means [&] ?

wadeliuyi commented on src/common/base/NebulaKeyUtils.cpp:
```diff
@@ -88,6 +88,59 @@ std::string NebulaKeyUtils::kvKey(PartitionID partId, const folly::StringPiece&
 
+// static
+void NebulaKeyUtils::indexRaw(const IndexValues &values, std::string& raw) {
+    std::vector<int32_t> colsLen;
```
colsLen init with size first better

wadeliuyi commented on src/storage/AddEdgesProcessor.cpp:
```diff
@@ -13,27 +13,133 @@ namespace nebula {
-                                               edge.key.ranking, edge.key.dst, version);
-            data.emplace_back(std::move(key), std::move(edge.get_props()));
+    if (indexes_.empty()) {
+        std::for_each(req.parts.begin(), req.parts.end(), [&](auto& partEdges){
```
no need capture all local variables

bright-starry-sky replied:
your means `[&]` ?

dangleptr commented on src/storage/BaseProcessor.inl:
```diff
@@ -83,5 +83,54 @@ void BaseProcessor<RESP>::doRemoveRange(GraphSpaceID spaceId,
+    auto&& v = value(std::move(res));
+    switch (col.get_type().get_type()) {
+        case nebula::cpp2::SupportedType::BOOL: {
+            return folly::to<std::string>(boost::get<bool >(v));
```
folly::to is not reasonable for encode bool

dangleptr commented on src/storage/BaseProcessor.inl:
```diff
@@ -83,5 +83,54 @@ void BaseProcessor<RESP>::doRemoveRange(GraphSpaceID spaceId,
+        case nebula::cpp2::SupportedType::TIMESTAMP: {
+            std::string raw;
+            raw.reserve(sizeof(int64_t));
+            auto val = folly::Endian::big(boost::get<int64_t>(v));
```
Take care of negative numbers.

dangleptr commented on src/storage/BaseProcessor.inl:
```diff
@@ -83,5 +83,54 @@ void BaseProcessor<RESP>::doRemoveRange(GraphSpaceID spaceId,
+        }
+        case nebula::cpp2::SupportedType::FLOAT:
+        case nebula::cpp2::SupportedType::DOUBLE: {
+            return folly::to<std::string>(boost::get<double>(v));
```
Don't use folly::to

dangleptr commented on src/interface/storage.thrift:
```diff
@@ -163,6 +163,12 @@ struct Edge {
 
+struct IndexItem {
+     1: required common.IndexID          index_id,
+     2: required i32                     schema,
```
Why "schema" type is i32? 

It seems the name should be tagIdOrEdgeType.

dangleptr replied:
A better name will be helpful when reading your code.  
So we'd better give a more meaningful name here.  
How about "tagOrEdge", short and meaningful.

bright-starry-sky replied:
> A better name will be helpful when reading your code.
> So we'd better give a more meaningful name here.
> How about "tagOrEdge", short and meaningful.

I agree with `tagOrEdge`

dangleptr commented on src/storage/mutate/AddEdgesProcessor.cpp:
```diff
@@ -13,27 +13,133 @@ namespace nebula {
+        auto srcId = edge.key.src;
+        auto rank = edge.key.ranking;
+        auto dstId = edge.key.dst;
+        VLOG(4) << "PartitionID: " << partId << ", VertexID: " << srcId
```
The highest vlog level is 3 in storage layer.

dangleptr commented on src/storage/mutate/AddEdgesProcessor.cpp:
```diff
@@ -13,27 +13,133 @@ namespace nebula {
+             * skip the in-edge.
+             **/
+            if (index.get_schema() == type) {
+                cpp2::EdgeKey ek(apache::thrift::FragileConstructor::FRAGILE,
```
Don't use "apache::thrift::FragileConstructor::FRAGILE" any more. It has been deleted in the latest fbthrift.

bright-starry-sky replied:
Good point ! acquire knowledge from this.

dangleptr commented on src/storage/mutate/AddEdgesProcessor.cpp:
```diff
@@ -13,27 +13,133 @@ namespace nebula {
+std::string AddEdgesProcessor::addEdges(int64_t version, PartitionID partId,
+                                        const std::vector<cpp2::Edge>& edges) {
+    std::unique_ptr<kvstore::BatchHolder> batchHolder = std::make_unique<kvstore::BatchHolder>();
+    std::map<std::string, std::pair<std::string,
```
Add more comments for this structure.  
It confused me.

dangleptr commented on src/storage/mutate/AddEdgesProcessor.h:
```diff
@@ -28,6 +29,19 @@ class AddEdgesProcessor : public BaseProcessor<cpp2::ExecResponse> {
+                              const std::pair<cpp2::EdgeKey, cpp2::IndexItem>& pair);
+
+    std::string newIndex(PartitionID partId,
+            const std::pair<std::string, std::pair<cpp2::EdgeKey, cpp2::IndexItem>>& index);
```
Alignment.

dangleptr commented on src/storage/mutate/AddEdgesProcessor.cpp:
```diff
@@ -13,27 +13,133 @@ namespace nebula {
+}
+
+std::string AddEdgesProcessor::newIndex(PartitionID partId,
+        const std::pair<std::string, std::pair<cpp2::EdgeKey, cpp2::IndexItem>>& index) {
```
Alignment

dangleptr commented on src/storage/mutate/AddEdgesProcessor.cpp:
```diff
@@ -13,27 +13,133 @@ namespace nebula {
+        VLOG(4) << "PartitionID: " << partId << ", VertexID: " << srcId
+                << ", EdgeType: " << type << ", EdgeRanking: " << rank
+                << ", VertexID: " << dstId << ", EdgeVersion: " << version;
+        auto key = NebulaKeyUtils::edgeKey(partId, srcId, type, rank, dstId, version);
```
When inserting an edge,   the operations looks like:
1.  Get the original value for the edgeKey.
2.  Delete the old index for the related indexId
3.  Insert the new index and data.

bright-starry-sky commented:
Addressed all comments since 2019-12-26 23:00:00
Also , encode `int64_t` and `double` for string order completed.

bright-starry-sky commented:
Jenkins go

dangleptr commented on src/storage/mutate/AddEdgesProcessor.cpp:
```diff
@@ -13,27 +13,163 @@ namespace nebula {
+        for (auto& index : iRet.value()) {
+            indexes_.emplace_back(index);
+        }
+    }
```
What if !iRet.ok(),  return error directly.

bright-starry-sky replied:
This schemaMan_ is going to be replaced by indexMan_ .
Currently, return` !iRet.ok()` means that the index does not exist

dangleptr commented on src/storage/mutate/AddEdgesProcessor.cpp:
```diff
@@ -13,27 +13,163 @@ namespace nebula {
+        /*
+         * step 2 , Insert new edge data
+         */
+        auto key = index.first;
```
auto&

bright-starry-sky replied:
> auto&

Aha, because the argument of method `batchHolder->put` is an rvalue reference ，Here I use std::move to cast the reference.  So `auto&` doesn't work.

dangleptr commented on src/storage/mutate/AddEdgesProcessor.cpp:
```diff
@@ -13,27 +13,163 @@ namespace nebula {
+         * step 2 , Insert new edge data
+         */
+        auto key = index.first;
+        auto prop = index.second.first;
```
auto&

dangleptr commented on src/storage/mutate/AddEdgesProcessor.cpp:
```diff
@@ -13,27 +13,163 @@ namespace nebula {
+                                                     NebulaKeyUtils::getDstId(rawKey),
+                                                     values);
+        std::string val;
+        auto result = kvstore_->get(spaceId_, partId, indexKey, &val);
```
What's the purpose about this get.

bright-starry-sky replied:
> What's the purpose about this get.

First, we need to check if there is an older version of this edge. If it exists, then index row corresponding to old version data must also exist. So this `get` is redundant. you are right. good point.

bright-starry-sky commented on src/storage/BaseProcessor.inl:
```diff
@@ -83,5 +83,54 @@ void BaseProcessor<RESP>::doRemoveRange(GraphSpaceID spaceId,
+
+template <typename RESP>
+IndexValues BaseProcessor<RESP>::collectIndexValues(RowReader* reader,
+    const std::vector<nebula::cpp2::ColumnDef>& cols) {
```
> Alignment

I was stumped 🤔, This parameter is too long. What should I do？

bright-starry-sky commented on src/storage/BaseProcessor.inl:
```diff
@@ -83,5 +83,54 @@ void BaseProcessor<RESP>::doRemoveRange(GraphSpaceID spaceId,
 }
 
+template <typename RESP>
+std::string BaseProcessor<RESP>::indexStr(RowReader* reader,
```
Of course, I'll do it.

dangleptr commented on src/storage/mutate/DeleteVertexProcessor.cpp:
```diff
@@ -17,54 +17,131 @@ void DeleteVertexProcessor::process(const cpp2::DeleteVertexRequest& req) {
+                                                               vId,
+                                                               values);
+                std::string val;
+                auto result = kvstore_->get(spaceId, partId, indexKey, &val);
```
We don't need the 'get' operation.

dangleptr commented on src/storage/mutate/DeleteVertexProcessor.cpp:
```diff
@@ -17,54 +17,131 @@ void DeleteVertexProcessor::process(const cpp2::DeleteVertexRequest& req) {
+     * so delete duplicate index keys is required.
+     **/
+    if (!indexes.empty()) {
+        sort(indexes.begin(), indexes.end());
```
We should skip the duplicate versions key when iterating.

bright-starry-sky replied:
Make sense!

bright-starry-sky commented on src/storage/mutate/UpdateEdgeProcessor.cpp:
```diff
@@ -260,11 +262,50 @@ std::string UpdateEdgeProcessor::updateAndWriteBack() {
+    if (!indexes_.empty()) {
+        for (auto& index : indexes_) {
+            auto indexId = index.get_index_id();
+            if (index.get_tagOrEdge() == edgeKey.edge_type) {
```
There is no possible, Because ID is unique whether it's TagId or EdgeType

bright-starry-sky commented:
Jenkins go

bright-starry-sky commented:
Jenkins go

dutor commented:
Jenkins go

critical27 commented on src/common/base/NebulaKeyUtils.h:
```diff
@@ -192,6 +214,68 @@ class NebulaKeyUtils final {
+     */
+
+    static std::string encodeInt64(int64_t v) {
+        v ^= folly::to<int64_t>(1) << 63;
```
How about define a constexpr?

bright-starry-sky replied:
A good idea, Do you mean for method `encodeInt64` or constant `folly::to<int64_t>(1) << 63`?

dangleptr commented on src/storage/mutate/AddEdgesProcessor.cpp:
```diff
@@ -13,27 +13,156 @@ namespace nebula {
     });
+    for (auto& e : newIEdges) {
+        for (auto& index : indexes_) {
+            if (NebulaKeyUtils::getEdgeType(e.first) == index.get_tagOrEdge()) {
```
1.  It seems that we don't need the if condition.  
2. We'd better get the original value in the first step before processing all indexes_

bright-starry-sky commented:
Performance optimization for index write . 
The performance report please refer to #1580

dangleptr commented on src/storage/mutate/AddEdgesProcessor.cpp:
```diff
@@ -13,27 +13,157 @@ namespace nebula {
+        std::string val;
+        std::unique_ptr<RowReader> nReader;
+        auto edgeType = NebulaKeyUtils::getEdgeType(e.first);
+        for (auto& index : indexes_) {
```
Preprocess the indexes_,   for each edgeType,  we just need to get the original value one time.

bright-starry-sky replied:
Yes, now is.

dangleptr replied:
Got it

dangleptr commented on src/storage/StorageFlags.cpp:
```diff
@@ -16,3 +16,5 @@ DEFINE_int32(waiting_catch_up_interval_in_secs, 30,
 DEFINE_int32(waiting_new_leader_retry_times, 30, "retry times when waiting for catching up data");
 DEFINE_int32(waiting_new_leader_interval_in_secs, 5,
              "interval between two requests for catching up state");
+DEFINE_bool(ignore_index_check_pre_insert, false,
```
Well done.

dangleptr commented on src/interface/common.thrift:
```diff
@@ -100,4 +102,10 @@ struct Pair {
 
+struct IndexItem {
+     1: required IndexID          index_id,
+     2: required i32              tagOrEdge,
```
By the way,   why we need tagOrEdge here?  index_id is not enough?

And where did we use the IndexItem?

bright-starry-sky replied:
Different index id may correspond to the same tagID.
IndexItem will be used for storage layer and graph lookup .

bright-starry-sky replied:
> By the way, why we need tagOrEdge here? index_id is not enough?
> 
> And where did we use the IndexItem?

I need a detailed explanation：
When addVertices or addEdges start executing, maybe have multiple kinds of tags or edgeType in the same request. And a tag or edgeType may have multiple indexes.Therefore, I need to know the relationship between specific tags and indexes.

Also, IndexItem will be used for each vertex or edge write, for example insert|delete|update ，the IndexId and tagOrEdge is required.

dangleptr replied:
Make sense.

darionyaphet commented on src/common/base/NebulaKeyUtils.cpp:
```diff
@@ -86,6 +86,59 @@ std::string NebulaKeyUtils::kvKey(PartitionID partId, const folly::StringPiece&
+        }
+        raw.append(col.second.data(), col.second.size());
+    }
+    for (auto len : colsLen) {
```
why append column size at the end of encoded-value?

bright-starry-sky replied:
Used for variable-length column type ， for example `string`  .
if an index row likes string('abc') + string('abc') , when the where expression is col1 == 'abca', the current row should be skip.

dangleptr commented on src/storage/mutate/UpdateVertexProcessor.cpp:
```diff
@@ -259,13 +261,41 @@ std::string UpdateVertexProcessor::updateAndWriteBack() {
+                                                              spaceId_,
+                                                              u.first);
+                    const auto &cols = index.get_cols();
+                    auto values = collectIndexValues(std::move(reader).get(), cols);
```
why std::move(reader) ?

bright-starry-sky replied:
Good point! It got me thinking about performance optimization. I've solved the performance problem for delete|update vertices or edges.

bright-starry-sky replied:
> why std::move(reader) ?

Good point! It got me thinking about performance optimization. I've solved the performance problem for delete|update vertices or edges.

dangleptr commented on src/storage/mutate/UpdateVertexProcessor.cpp:
```diff
@@ -259,13 +261,41 @@ std::string UpdateVertexProcessor::updateAndWriteBack() {
+                                                                   spaceId_,
+                                                                   u.first);
+                        const auto &oCols = index.get_cols();
+                        auto oValues = collectIndexValues(std::move(oReader).get(), oCols);
```
ditto

bright-starry-sky commented:
1,Performance optimization. 2,Resolve the conflict. 3,Improve test cases

jude-zhu commented:
close #458
